src/pyggdrasil/tree_inference/_simulate.py

```diff
@@ -44,15 +47,58 @@ def add_noise_to_perfect_matrix(
             - ``MISSING ENTRY`` if ``missing_entry_rate`` is non-zero
     """
     # RNGs for false positives, false negatives, and missing data
-    rng_false_pos, rng_false_neg, rng_miss = random.split(rng, 3)
-
-    # TODO(Pawel, Gordon): Now probably it's the easiest to draw from random.bernoulli
-    #   matrices corresponding to the false positives and false negatives
-    #   and adjust the mutation matrix accordingly
-
-    # Now the matrix can be adjusted according to missing entry simulation
-
-    raise NotImplementedError("This function needs to be implemented.")
+    rng_false_pos, rng_false_neg, rng_miss, rng_homo_pos, rng_homo_neg = random.split(
```

Comment:
Very nice you did it! However, I'd encourage you to split the body of the function into several smaller functions (a function with 50 lines of code, multiple steps, involving complex constructions may be hard to understand in a month or two).

Reply:
Alright, split add_noise_to_perfect_matrix() into four private functions.